File: Search/BinarySearch/BinarySearch.cpp
/*
* Time Complexity: O(log N)
 *
 * Auxiliary Space: O(1)
 *
 * !ONLY FOR SORTED DATA ARRAYS!
 */

#include <iostream>
#include <cinttypes>

int main() {
    size_t n;
    std::cin >> n;

    auto* data = new int64_t[n];
    for (size_t i = 0; i < n; ++i) {
        std::cin >> data[i];
    }

    int64_t find_value;
    std::cin >> find_value;

    size_t value_index = -1;
    size_t left = 0;
    size_t right = n;
    while (left + 1 < right) {
        size_t mid = (left + right) / 2;

        if (data[mid] == find_value) {
            value_index = mid;
            break;
        }

        if (data[mid] <= value_index) {
            left = mid;
        } else {
            right = mid;
        }
    }

    std::cout << value_index << std::endl;

    delete[] data;
}

File: Search/BinarySearch/BinarySearch.kt
/*
* Time Complexity: O(log N)
 *
 * Auxiliary Space: O(1)
 *
 * !ONLY FOR SORTED DATA ARRAYS!
 */

import kotlin.random.Random

fun<T> Collection<T>.printList() {
    this.forEach {
        print("$it ")
    }
}

fun main() {
    val n : Int = Random.nextInt(1, 100000)
    val data : MutableList<Int> = MutableList(n) { Random.nextInt(Int.MIN_VALUE, Int.MAX_VALUE) }

    data.sort()

    val key : Int = data[Random.nextInt(0, n)]
    var position : Int = -1

    var left : Int = 0
    var right : Int = n

    while (left + 1 < right) {
        val mid : Int = (left + right) / 2

        if (data[mid] == key) {
            position = mid
            break
        }

        if (data[mid] <= key) {
            left = mid
        } else {
            right = mid
        }
    }

    position = when (position) {
        -1 -> when (data[left]) {
            key -> left
            else -> -1
        }
        else -> position
    }

    println("Target value is $key")
    println("Target value position in data is $position")
}
